kerchunk/grib2.py

```diff
@@ -100,148 +97,133 @@ def scan_grib(
+    common_vars: (depr, do not use)
     storage_options: dict
         For accessing the data, passed to filesystem
-    inline_threashold: int
+    inline_threshold: int
         If given, store array data smaller than this value directly in the output
     skip: int
         If non-zero, stop processing the file after this many messages
     filter: dict
-        cfgrib-style filter dictionary
+        keyword filtering. For each key, only messages where the key exists and has
+        the exact value or is in the given set, are processed.
+        E.g., the cf-style filter ``{'typeOfLevel': 'heightAboveGround', 'level': 2}``
+        only keeps messages where heightAboveGround==2.
 
     Returns
     -------
 
-    dict: references dict in Version 1 format.
+    list(dict): references dicts in Version 1 format, one per message in the file
     """
-    common_vars = common_vars or []
     storage_options = storage_options or {}
-    if filter:
-        assert "typeOfLevel" in filter
     logger.debug(f"Open {url}")
 
-    store = {}
-    z = zarr.open_group(store, mode="w")
-    common = False
+    out = []
     with fsspec.open(url, "rb", **storage_options) as f:
-        for fn, offset, size in _split_file(f, skip=skip):
-            logger.debug(f"File {fn}")
-            ds = cfgrib.open_file(fn)
-            if filter:
-                var = filter["typeOfLevel"]
-                if var not in ds.variables:
+        logger.debug(f"File {url}")
+        for offset, size, data in _split_file(f, skip=skip):
+            store = {}
+            mid = eccodes.codes_new_from_message(data)
+            m = cfgrib.cfmessage.CfMessage(mid)
+
+            good = True
+            for k, v in (filter or {}).items():
+                if k not in m:
+                    good = False
+                elif isinstance(v, (list, tuple, set)):
+                    if m[k] not in v:
+                        good = False
+                elif m[k] != v:
+                    good = False
+            if good is False:
+                continue
+
+            z = zarr.open_group(store)
+            global_attrs = {
+                k: m[k] for k in cfgrib.dataset.GLOBAL_ATTRIBUTES_KEYS if k in m
+            }
+            z.attrs.update(global_attrs)
+
+            vals = m["values"].reshape((m["Ny"], m["Nx"]))
+            attrs = {
+                k: m[k]
+                for k in cfgrib.dataset.DATA_ATTRIBUTES_KEYS
+                + cfgrib.dataset.DATA_TIME_KEYS
+                + cfgrib.dataset.EXTRA_DATA_ATTRIBUTES_KEYS
+                if k in m
+            }
+            _store_array(
+                store, z, vals, m["shortName"], inline_threshold, offset, size, attrs
+            )
+            if "typeOfLevel" in m and "level" in m:
+                name = m["typeOfLevel"]
+                data = np.array([m["level"]])
+                attrs = cfgrib.dataset.COORD_ATTRS[name]
```

Comment:
@martindurant Updating after release 0.0.7 I am now getting the following error using scan_grib with HRRR: gcs://high-resolution-rapid-refresh/hrrr.20220720/conus/hrrr.t00z.wrfsfcf21.grib2

```
zarr_meta = scan_grib(input_path, **SCAN_SURFACE_GRIB)
  File "/Users/dstuebe/.cache/bazel/_bazel_dstuebe/76bbf57da584b86027104686797623fa/execroot/ritta/bazel-out/k8-dbg/bin/ingestion/noaa_nwp/run_local.runfiles/common_deps_kerchunk/kerchunk/grib2.py", line 160, in scan_grib
    attrs = cfgrib.dataset.COORD_ATTRS[name]
KeyError: 'surface'
```
where
```
SCAN_SURFACE_GRIB = dict(
    common=["time", "step", "latitude", "longitude", "valid_time"],
    filter=dict(stepType="instant", typeOfLevel="surface"),
    storage_options=dict(token=None),
)
````
I am afraid I am a bit lost on how to start debugging this and wishing I tested last week. I tried shuffling around my filter keywords a bit but I can't grok how this changed from the previous iteration?

Reply:
I had meant to merge https://github.com/fsspec/kerchunk/pull/204 before release... Can you try with that? I believe it runs, but note that the grib scanner no longer does any merging by itself, but creates a list of output dicts that you need to pass to MultiZarrToZarr. With that PR, I think you will still get a coordinate for the typeOfLevel parameter, which you might argue would be better as a dataset attribute. Also, I think I see a bug (which won't stop you, but the attributes of that one-value coordinate variable will be wrong), so maybe it's just as well I didn't merge yet.

Reply:
This works - thank you.
Will look forward to the release 0.0.8 when you have time.